### PR TITLE
FIX KappaTune does not dequantize bitsandbytes 8-bit weights

### LINKED ISSUE #3736: KappaTuneSelector does not dequantize bitsandbytes 8-bit weights, computing condition numbers on raw int8 codes
### System Info

- peft: main (`0e8d0ae8`)
- transformers: 5.16.1
- torch: 2.13.0
- Python 3.11
- Affects bitsandbytes 8-bit models (`load_in_8bit=True`); independent of platform

### Who can help?

@benjaminbossan @githubnemo

### Reproduction

`KappaTuneSelector._compute_kappas` tries to dequantize bnb 8-bit weights, but it reads `CB`/`SCB` from the wrong object:

https://github.com/huggingface/peft/blob/0e8d0ae8ab94f189f28b845e293d7452d7892d91/src/peft/helpers.py#L446-L455

```python
weight = module.weight
if bnb is not None:
    if hasattr(weight, "quant_state"):  # 4-bit
        w = bnb.functional.dequantize_4bit(weight.data, weight.quant_state).float()
    elif hasattr(weight, "state") and hasattr(weight.state, "CB"):  # int8
        w = bnb.functional.int8_vectorwise_dequant(weight.state.CB, weight.state.SCB).float()
    else:
        w = weight.data.detach().float()
```

In bitsandbytes, `CB` and `SCB` are attributes of the `Int8Params` weight itself (`self.CB = CB`, `self.SCB = SCB` in `bitsandbytes/nn/modules.py`). `state` (`MatmulLtState`) belongs to the `Linear8bitLt` *module*, not to its weight. So `hasattr(weight, "state")` is `False` for 8-bit weights, and the int8 branch is never taken.

The code then falls through to `w = weight.data.detach().float()`, which computes the SVD on the raw int8 codes without applying the per-row scales. Because each row is scaled by a different factor, the resulting condition numbers (and therefore the selected targets) differ from those of the actual weights. This happens silently, with no error or warning.

For comparison, PEFT's own `dequantize_bnb_weight` in `src/peft/utils/integrations.py` handles 8-bit correctly by taking the module's `state` and falling back to `weight.SCB`.

The existing GPU test `test_kappatune_with_4bit_model` in `tests/test_gpu_examples.py` only covers 4-bit, so this path isn't exercised.

Note: I found this by reading the PEFT and bitsandbytes source; I don't have a GPU available to run it. A reproducer along these lines should show the problem (untested):

```python
import torch
from transformers import AutoModelForCausalLM, BitsAndBytesConfig
from peft.helpers import KappaTuneSelector

model_id = "facebook/opt-125m"

model_8bit = AutoModelForCausalLM.from_pretrained(
    model_id, quantization_config=BitsAndBytesConfig(load_in_8bit=True), device_map="cuda"
)
module = model_8bit.model.decoder.layers[0].self_attn.q_proj
print(type(module.weight).__name__, hasattr(module.weight, "state"), hasattr(module.weight, "CB"))
# expected: Int8Params False True  -> int8 branch is skipped

model_fp = AutoModelForCausalLM.from_pretrained(model_id, torch_dtype=torch.float32, device_map="cuda")

kappa_8bit = KappaTuneSelector(model_8bit, show_progress=False)
kappa_fp = KappaTuneSelector(model_fp, show_progress=False)
kappa_8bit._compute_kappas()
kappa_fp._compute_kappas()

name = "model.decoder.layers.0.self_attn.q_proj"
print(kappa_8bit._condition_numbers[name], kappa_fp._condition_numbers[name])
# expected: values differ noticeably, since the 8-bit one is computed on unscaled int8 codes
```

### Expected behavior

For 8-bit bnb models, KappaTune should dequantize the weights (e.g. via `weight.CB`/`weight.SCB`, or by reusing `dequantize_bnb_weight(module.weight, state=module.state)`) before computing condition numbers. Its target selection should then approximately match the selection for the same model loaded unquantized. An 8-bit case could be added next to `test_kappatune_with_4bit_model`.

### PR DESCRIPTION
**Pull request by amogh-nagri-11**
## What does this PR do?

Fixes #3736. Approved by @BenjaminBossan in https://github.com/huggingface/peft/issues/3736#issuecomment-5697944828.

`KappaTuneSelector._compute_kappas` read `CB`/`SCB` from `weight.state`, but for bitsandbytes `Int8Params` these attributes live on the weight itself, while `state` belongs to the `Linear8bitLt` module. The 8-bit branch was therefore never taken and condition numbers were computed on the raw int8 values, without their per-row scales. Since 8-bit quantization stores one scale per row, this can change the condition number by orders of magnitude and leads to different `target_modules` being selected.

## Changes

- Replace the hand-written 4-bit/8-bit branching with the existing `dequantize_module_weight` helper from `peft.utils.integrations`, which handles both bnb dtypes and returns unquantized weights unchanged.
- Drop the `try: import bitsandbytes` block in `helpers.py`, which was only used by that branching.

## Tests

Added `test_kappatune_with_8bit_model` to `tests/test_gpu_examples.py`, next to the existing 4-bit test and with the same markers. It rescales the rows of each linear weight of `tiny-random-LlamaForCausalLM`, loads the model in full precision and in 8-bit, and asserts that each layer's condition numbers agree within a factor of 2. The rescaling is needed because the tiny random weights have similar row scales, which masks the bug; the loose bound is needed because 8-bit rounding noise alone shifts the condition number of the nearly singular 16x16 `v_proj` by about 47%.

Without the fix the condition numbers are about 4.5x too small in the best case and about 150x too small for the median layer, so the test fails on `main` and passes with the change. The existing `test_kappatune_with_4bit_model` still passes.

I don't have a CUDA GPU, so I ran the test on CPU and on MPS with `PYTORCH_ENABLE_MPS_FALLBACK=1`; it needs a CI runner for real 8-bit coverage.

## AI assistance

AI assistance (Claude Code) was used for this change. I reviewed every changed line, ran the tests listed above, and can explain and defend the change.

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/peft/blob/main/CONTRIBUTING.md)?
- [x] Did you make sure to update the tests?

### CODE REVIEW

File: tests/test_gpu_examples.py [hunk 1]
```diff
@@ -7199,3 +7126,42 @@ def test_kappatune_with_4bit_model():
...
+    selector_8bit = KappaTuneSelector(model_8bit, show_progress=False)
+    selector_8bit._compute_kappas()
+
+    kappas_fp = selector_fp._condition_numbers
+    kappas_8bit = selector_8bit._condition_numbers
+    assert kappas_8bit.keys() == kappas_fp.keys()
+    # Quantization noise can still noticeably change the condition number of nearly singular weights, hence the loose
+    # bound. Without dequantization, the condition numbers would be off by a much larger factor.
+    for name, kappa_fp in kappas_fp.items():
+        assert 0.5 < kappas_8bit[name] / kappa_fp < 2, name
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py]:
```suggestion
        assert 0.5 < kappas_8bit[name] / kappa_fp < 2
```

No need, but if you want to keep it, make it a real message, just the name is confusing.

File: tests/test_gpu_examples.py [hunk 2]
```diff
@@ -7199,3 +7126,42 @@ def test_kappatune_with_4bit_model():
     assert "target_modules" in targets
     assert isinstance(targets["target_modules"], list)
     assert len(targets["target_modules"]) > 0, "Should return at least some target modules"
+
+
+@pytest.mark.single_gpu_tests
+@require_bitsandbytes
+def test_kappatune_with_8bit_model(tmp_path):
+    """Test that KappaTune dequantizes 8-bit quantized weights before computing condition numbers."""
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py]: I think the docstring doesn't quite explain this test. You would think that if the 8bit weight is not dequantized, we would get an error, so the same test approach as above would work. Let's update the explanation and link the issue for further info.

File: tests/test_gpu_examples.py [hunk 3]
```diff
@@ -7199,3 +7126,42 @@ def test_kappatune_with_4bit_model():
...
+    import torch
+    from transformers import AutoModelForCausalLM, BitsAndBytesConfig
+
+    from peft.helpers import KappaTuneSelector
+
+    model = AutoModelForCausalLM.from_pretrained("hf-internal-testing/tiny-random-LlamaForCausalLM")
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py]: Please move this call into the `hub_online_once` context. Let's also update the test above to do the same.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for providing this fix, good idea to use the `dequantize_module_weight` function here. There are some comments regarding the test, otherwise the PR looks good.

File: tests/test_gpu_examples.py (hunk 1 above)
Comment (amogh-nagri-11) [reply tests/test_gpu_examples.py]: Done, dropped the message.

File: tests/test_gpu_examples.py (hunk 2 above)
Comment (amogh-nagri-11) [reply tests/test_gpu_examples.py]: You're right, the point wasn't clear. Expanded the docstring to say that reading the int8 values without their per-row scales doesn't raise but silently gives condition numbers that are orders of magnitude off, which is why they're compared against the full precision model, and linked #3736.

File: tests/test_gpu_examples.py (hunk 3 above)
Comment (amogh-nagri-11) [reply tests/test_gpu_examples.py]: Done, both tests now load the model inside `hub_online_once`.

File: tests/test_gpu_examples.py [hunk 4]
```diff
@@ -7199,3 +7128,48 @@ def test_kappatune_with_4bit_model():
...
+
+@pytest.mark.single_gpu_tests
+@require_bitsandbytes
+def test_kappatune_with_8bit_model(tmp_path):
+    """Test that KappaTune dequantizes 8-bit quantized weights before computing condition numbers.
+
+    Reading the int8 values without their per-row scales does not raise, it silently returns condition numbers that are
+    orders of magnitude off, so they are compared to those of the same model in full precision. See #3736.
+    """
+    import torch
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py]: Let's make the imports global and remove them here, as well as in the other KappaTune test.

File: tests/test_gpu_examples.py [hunk 5]
```diff
@@ -7199,3 +7128,48 @@ def test_kappatune_with_4bit_model():
     assert "target_modules" in targets
     assert isinstance(targets["target_modules"], list)
     assert len(targets["target_modules"]) > 0, "Should return at least some target modules"
+
+
+@pytest.mark.single_gpu_tests
+@require_bitsandbytes
+def test_kappatune_with_8bit_model(tmp_path):
```
Comment (BenjaminBossan) [inline tests/test_gpu_examples.py]: Let's move this test and the other KappaTune test into a test class for this method.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the update. There are two more things I missed on my first review, please check.

File: tests/test_gpu_examples.py (hunk 4 above)
Comment (amogh-nagri-11) [reply tests/test_gpu_examples.py]: Done, the imports are now at module level for both tests.

File: tests/test_gpu_examples.py (hunk 5 above)
Comment (amogh-nagri-11) [reply tests/test_gpu_examples.py]: Done, both tests are now in `TestKappaTune`, with the markers moved to the class.

**Review by BenjaminBossan — APPROVED** [review]
Thanks for fixing this, LGTM.
